src/crypto/public_key/elliptic/ecdsa.js

```diff
@@ -25,25 +25,52 @@
 
 'use strict';
 
+import BN from 'bn.js';
+import ASN1 from 'asn1.js';
+import jwkToPem from 'jwk-to-pem';
+
 import curves from './curves.js';
 import BigInteger from '../jsbn.js';
+import config from '../../../config';
+import enums from '../../../enums.js';
+import util from '../../../util.js';
+import base64 from '../../../encoding/base64.js';
+
+const webCrypto = util.getWebCrypto();
+const webCurves = curves.webCurves;
+const nodeCrypto = util.getNodeCrypto();
+const nodeCurves = curves.nodeCurves;
+
+var ECDSASignature = ASN1.define('ECDSASignature', function() {
+  this.seq().obj(
+    this.key('r').int(),  // FIXME int or BN?
+    this.key('s').int()   // FIXME int or BN?
+  );
+});
 
 /**
  * Sign a message using the provided key
  * @param  {String}      oid        Elliptic curve for the key
- * @param  {BigInteger}  Q          Public key used to sign
  * @param  {enums.hash}  hash_algo  Hash algorithm used to sign
  * @param  {Uint8Array}  m          Message to sign
-* @param  {BigInteger}  d          Private key used to sign
+ * @param  {BigInteger}  d          Private key used to sign
  * @return {{r: BigInteger, s: BigInteger}}  Signature of the message
  */
-function sign(oid, hash_algo, m, d) {
+async function sign(oid, hash_algo, m, d) {
+  var signature;
   const curve = curves.get(oid);
+  hash_algo = hash_algo ? hash_algo : curve.hash;
   const key = curve.keyFromPrivate(d.toByteArray());
-  const signature = key.sign(m, hash_algo);
+  if (webCrypto && config.use_native && curve.web) {
+    signature = await webSign(curve, hash_algo, m, key.keyPair);
+  } else if (nodeCrypto && config.use_native && curve.node) {
+    signature = await nodeSign(curve, hash_algo, m, key.keyPair);
+  } else {
+    signature = await key.sign(m, hash_algo);
+  }
```

Comment:
One of the intentions while splitting KeyPair and Curve was to later use the Facade pattern. So you do not have to use lots of `if`s, the `curve.get()` and `curve.keyFromPrivate()` will instantiate the appropiate objects NodeCurve, WebcryptoCurve, PlainJSCurve, NodeKeyPair, WebCryptKeyPair, etc.

Reply:
@mahrud we can stuff webSign and nodeSign into the key.Sign function to achieve this

Reply:
Ahh, now I understand. Ok, I'll incorporate this by moving the fallback pattern into `key.js`.